CONTRIBUTING.md

```diff
@@ -139,7 +139,7 @@ As languages differ quite a lot, here is a set of captures available to you when
 @type             ; type or class definitions and annotations
 @type.builtin     ; built-in types
 @type.definition  ; identifiers in type definitions (e.g. `typedef <type> <identifier>` in C)
-@type.qualifier   ; type qualifiers (e.g. `const`)
+@type.qualifier   ; type qualifiers that are not language keywords
```

Comment:
could you add a concrete example for that?

Reply:
I don't understand what the 3 remaining `@type.qualifier` are for, but anyway Rust lifetime would be a good example. In that case I'd like to add a separate commit to change rust lifetime to `@type.modifier` in this PR (instead of a dedicated PR).

Reply:
:thinking: But at least from rustlang doc, `'static` is still a keyword, just not reserved keywords 

https://doc.rust-lang.org/reference/keywords.html

Reply:
OK, if we can use it for Rust, it must be important ;) I'd recommend two commits:
1. adding/using `@keyword.modifier`
2. renaming/expanding `@type.modifier`

Reply:
I'd rather not have *another* modifier capture. It's fine to use `@keyword.*` for not-exactly-keywords sometimes.

Reply:
Yes, if `@keyword.storage` fits well enough, no sense in adding more.

But really, this is starting to look like we need to address this mess holistically by reconsidering _all_ captures and queries that are somewhat related. I'm not too keen on that, to be honest.

And I stand by my point: Removing captures is not (easily) reversible, so if in doubt, leave it in (just renaming is fine).

Reply:
> I'd rather not have another modifier capture. It's fine to use `@keyword.*` for not-exactly-keywords sometimes.

I am strongly against this. <sub>Specifically, the recent commit that removed the `.lifetime` subcapture that differentiates rust lifetimes from keywords like `const` is what triggered me to raise the issue and PR.</sub>

> Yes, if `@keyword.storage` fits well enough, no sense in adding more.

I noticed that `@keyword.storage` is used for things like `public`, `const`, etc, which basically overlaps the proposed role of `@keyword.modifier`. Even the CONTRIBUTING.md says it's for "modifiers that affect storage in memory or life-time". I'm fine with either of them as long as non-keywords don't fall into `@keyword.*`.

So here's a yet another proposal(s)
* Option 1: If the maintainers want to stick to exising new captures
    * Commit 1: Replace `@type.qualifier` used for language keyword with `@keyword.storage`
    * Commit 2: For rust lifetimes
        * use `@keyword.storage` for `'static` and `'_`
        * use `@type.qualifier` for others
* Option 2: If the maintainers don't mind introducing a new capture
    * Commit 1: Replace `@type.qualifier` used for language keyword with `@keyword.modifier`
    * Commit 2: Replace remaining `@type.qualifier` with `@type.modifier`
    * Commit 3: Replace `@keyword.storage` with `@keyword.modifier`
    * Commit 4: For rust lifetimes
        * use `@keyword.modifier` for `'static` and `'_`
        * use `@type.modifier` for others

Reply:
To be clear, we have to walk a line here between expressiveness for any given language and consistency across languages; these are directly competing goals. This project has decided to prioritize the second wherever they conflict.

In the past, we used to be permissive with captures (only enforce top-level captures so every capture is guaranteed to be highlighted _somehow_), but we realized that this lead to us missing too many inconsistencies (see above re: goal 2). So we locked things down so only the explicitly listed captures are valid; these are documented in Neovim so can be considered stable.

That being said, this doesn't preclude adding new ones provided they are **generally useful**; captures that only make sense for a single language are too much and better defined in a standalone filetype plugin. (This is similar to how Vim treats syntax files.) The burden of proof here rests on the proposer: if you can demonstrate (in a PR) that the capture can be used consistently in multiple languages, I'm inclined to accept it.

Reply:
See also https://github.com/nvim-treesitter/nvim-treesitter/issues/6205, which this should align with.

Reply:
After reviewing some more queries, it seems there is no meaningful difference between the current use of `@keyword.storage` and the proposed role of `@keyword.modifier`, and generally `@keyword.modifier` fits better due to its loose definition. So I'm inclined to the Option 2 (pro-`@keyword.modifier`).

But I have a very minor concern regarding `@keyword.modifier`. The term "modifier" has a connotation that it's optional, yet some keywords that fall under `@keyword.modifier` may not be optional in some cases. The `const` keyword in Rust is a good example.

- [const function](https://doc.rust-lang.org/reference/const_eval.html?highlight=const#const-functions): `const` is optional, so `@keyword.modifier` fits.
- [const item](https://doc.rust-lang.org/reference/items/constant-items.html#constant-items): Without `const`, it's a syntax error. Maybe `@keyword.storage` fits better?
- [`*const T` pointer](https://doc.rust-lang.org/reference/types/pointer.html?highlight=const#raw-pointers-const-and-mut): Without `const`, it's a syntax error because types cannot have the form of `* T`.
- [const generic parameter](https://doc.rust-lang.org/reference/items/generics.html?highlight=const#generic-parameters): Without `const`, it may or may not be a syntax error, because it might be a syntactically valid type parameter with type bound.

Addressing this issue would require writing more complex queries. I'm not really interested in that and I think it's fine to ignore it.

Reply:
The question is more about guidance:
1. when would query authors use one over the other?
2. how should colorschemes color one or the other?

(Remember these are _highlight_ queries, so that's the goal here. Coloring everything differently does not add value, so it's OK to -- meaningfully -- group things even if it's semantically not fully correct.)

Concretely, I would **not** let

> The term "modifier" has a connotation that it's optional

(which is a connotation I don't see, personally) be a blocker here; this can be handled by documentation. (For me, a "modifier" is simply something that doesn't make sense in itself but only in connection with a "base" -- which `const` fits in every case, doesn't it?)

Reply:
(Just to be clear: I don't want to block anything here; your initiative is very welcome. I just want to find the best long-time viable strategy, given the project goals.)

Reply:
## current uses of `@type.qualifier`

I'm trying to write a description for the non-keyword `@type.qualifier`, but its uses and my proposal to apply it to rust lifetime don't seem to be consistent.

### blueprint
https://github.com/nvim-treesitter/nvim-treesitter/blob/f8d4e5c1ba2b24b7d13830eeca394fc24c808d83/queries/blueprint/highlights.scm#L52-L53

Example: `Gtk.HeaderBar` in the following.
```
template $MapsHeaderBar : Gtk.HeaderBar {
  /* probably a lot of buttons ... */
}
```
(from <https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/reference/templates.html>)

`Gtk.HeaderBar` looks like some sort of inheritance base. `Gtk.HeaderBar` itself would be a valid type. Queries for other language don't specifically highlight inheritance base (C++, Java) or something similar (Rust trait bound).

#### chatito

<https://github.com/nvim-treesitter/nvim-treesitter/blob/f8d4e5c1ba2b24b7d13830eeca394fc24c808d83/queries/chatito/highlights.scm#L33>

Example: `#time_in_hours` and `#relative_time` in the following.
```
%[ask_for_delivery]
    my parcel should be delivered in @[delivery_time#time_in_hours]
    my parcel should be delivered @[delivery_time#relative_time]

@[delivery_time#time_in_hours]
    3 days
    5 hours

@[delivery_time#relative_time]
    as fast as possible
    quickly
```
(from <https://github.com/rodrigopivi/Chatito/blob/master/spec.md#212---slot>)

They modifiy the base `delivery_time`, and do not make sense without the base. This is quite different from blueprint's use.

### eds
<https://github.com/nvim-treesitter/nvim-treesitter/blob/f8d4e5c1ba2b24b7d13830eeca394fc24c808d83/queries/eds/highlights.scm#L40-L45>
This filetype is just configurations with [simple list of key/values](https://github.com/uyha/tree-sitter-eds/blob/main/grammar.js). It doesn't have much syntactic variations, so it seems to use random capture names for some specific built-in keys. I think we can ignore this case.

## what to do with `@type.qualifier`
I think this group can be expanded to other languages or removed.
* Expand: Declare that `@type.qualifier` is for things that are types themselves. Apply `@type.qualifier` for inheritance base and type bounds in the other languages.
* Remove: Don't specially treat inheritance base and type bounds.

In both cases, we are left with chatito's variation and rust lifetime. I can think of two solutions that don't involve inventing a new capture name.
* `@attribute`: This is used for things decorating existing stuff, so kinda makes sense? It also has `@attribute.builtin` subcapture, which `'static` and `'_` may fit. Also chatito's variation seems to fit.
* Alternatively for rust lifetime, use `@label`: Queries in some other project highlights rust lifetime as `@label` ([Helix](https://github.com/helix-editor/helix/blob/fd89c3c8335399e344e038f1141ea0657653a591/runtime/queries/rust/highlights.scm#L51-L56), [tree-sitter-rust](https://github.com/tree-sitter/tree-sitter-rust/blob/2d7bac5164c344c95530c16dd5c8fa1319e34308/queries/highlights.scm#L98)) (not in [Zed](https://github.com/zed-industries/zed/blob/main/crates/languages/src/rust/highlights.scm)). This seems to be an ok-ish choice as it looks the same as loop labels.

## other notes
* verilog has [`lifetime`](https://github.com/tree-sitter/tree-sitter-verilog/blob/2dfddfcb891f74a1cdb9d8ce480a997d11846e17/grammar.js#L1282), but it's highlighted `@label` in nvim-treesitter. It should be `@keyword.modifier`.
* lalrpop has lifetime, but it's a parser generator for rust, so it's basically Rust.

Reply:
Current usage is not consistent (for historical reasons). I'm not against changing things up if that is required for consistency. Agree that `@type.modifier` makes sense for chatito but not for blueprint.

(Also, blueprint and chatito are esolangs, feel free to adjust as needed for Rust.)